pandas/_libs/tslib.pyx

```diff
@@ -459,6 +461,55 @@ def array_with_unit_to_datetime(ndarray values, object unit,
     return oresult
 
 
+cdef get_key(tz):
+    if tz is None:
+        return None
+    if is_fixed_offset(tz):
+        # TODO: these should all be mapped together
+        try:
+            # pytz
+            return str(tz._minutes)  # pytz specific?
+        except AttributeError:
+            try:
+                # dateutil.tz.tzoffset
+                return str(tz._offset.total_seconds())
+            except AttributeError:
+                return str(tz)
+    return tz_cache_key(tz)
+
+
+cdef fixed_offset_to_pytz(tz):
```

Comment:
As mentioned in my comment, I don't think we shouldn't be converting dateutil objects to pytz objects unless completely necessary.

Reply:
It may be worth separately considering `pd.to_datetime(...)` vs `pd.DatetimeIndex(...)`.  The latter has a much stronger case for coercing equivalent tzinfos.  For the former, mixing and matching is sufficiently weird that the user might be doing it intentionally.

Also note that the existing code coerces dateutil tzoffsets parsed from strings to pytz.FixedOffset.  As a result, `to_datetime(x)` and `to_datetime(Timestamp(x))` can have different tzinfos.  Not _quite_ the same situation, but similar.

Reply:
>coercing equivalent tzinfos

If we end up doing this then, I would prefer to pick a consistent timezone object (probably pytz) and not pick a psudo-random one (e.g. pick the last one).

In the case of mixed timezones though, would we coerce to object or raise?

Reply:
>  I would prefer to pick a consistent timezone object (probably pytz) and not pick a psudo-random one (e.g. pick the last one).

So we would need some kind of hierarchy like:

```
def choose_utc(utcs):
    if pytz.utc in utcs:
        return pytz.utc
    elif PY3 and datetime.timezone.utc in utcs:
        return datetime.timezone.utc
    elif ...

def choose_fixed_offset(fixed):
    ....
```

> In the case of mixed timezones though, would we coerce to object or raise?

The approach I took in #23675 is to have `objects_to_dt64ns` have an `allow_object` kwarg that determines this.

Reply:
Another (far in the future) option is to use a pandas-equivalent tzinfo as discussed here: https://github.com/pandas-dev/pandas/issues/23959#issuecomment-443886222. But yeah I hierarchy would be needed in the short term.

 Since `objects_to_dt64ns` is just used internally, in what instances are we choosing for `allow_object` to be True or False?